### PR TITLE
use animation frames for memory autoplay

### PR DESCRIPTION
**Pull request by uhthomas**
The current implementation mixes intervals and animation frames, which is a little convoluted. The use of intervals means that the animation is not going to be smooth and may have strange behaviour when the window is moved to the background. It's possible that the current animation frames could pile up and run all at once which would be undesirable.

Moving everything into animation frames means the code is simpler and easier to reason about. It should also be more performant and less buggy.

### CODE REVIEW

**Comment by uhthomas** [comment]
I've tested and it works pretty well, but resumption seems broken. I guess I should fix that.

File: web/src/lib/components/memory-page/memory-viewer.svelte [hunk 1]
```diff
@@ -42,119 +45,108 @@
...
+			if (isNaN(memoryIndex) || memoryIndex > $memoryStore.length - 1) {
+				memoryIndex = 0;
 			}
 		}
+	});
 
-		currentMemory = $memoryStore[currentIndex];
+	onDestroy(async () => {
+		if (browser) pause();
+	});
```
Comment (jrasm91) [inline web/src/lib/components/memory-page/memory-viewer.svelte:57]:
```suggestion
	onDestroy(() => {
		if (browser) {
		  pause();
		}  
	});
```

File: web/src/lib/components/memory-page/memory-viewer.svelte [hunk 2]
```diff
@@ -42,119 +45,108 @@
...
 
-		nextIndex = currentIndex + 1;
-		nextMemory = $memoryStore[nextIndex];
+	const canAdvance = () => !!(nextMemory || nextAsset);
```
Comment (jrasm91) [inline web/src/lib/components/memory-page/memory-viewer.svelte:59]:
```suggestion
	$: canAdvance = !!(nextMemory || nextAsset);
```

File: web/src/lib/components/memory-page/memory-viewer.svelte [hunk 3]
```diff
@@ -246,29 +239,33 @@
...
 								</div>
 							</div>
 							<div class="flex h-full flex-col place-content-center place-items-center mr-4">
 								<div class="inline-block">
-									<CircleIconButton
-										logo={ChevronRight}
-										backgroundColor="#202123"
-										on:click={toNextCurrentAsset}
-									/>
+									{#if nextMemory || assetIndex < currentMemory.assets.length - 1}
```
Comment (jrasm91) [inline web/src/lib/components/memory-page/memory-viewer.svelte:253]: I think you can use `nextMemory || nextAsset` here.

File: web/src/lib/components/memory-page/memory-viewer.svelte [hunk 4]
```diff
@@ -246,29 +239,33 @@
 						<div class="absolute h-full flex justify-between w-full">
 							<div class="flex h-full flex-col place-content-center place-items-center ml-4">
 								<div class="inline-block">
-									<CircleIconButton
-										logo={ChevronLeft}
-										backgroundColor="#202123"
-										on:click={toPreviousCurrentAsset}
-									/>
+									{#if previousMemory || assetIndex > 0}
```
Comment (jrasm91) [inline web/src/lib/components/memory-page/memory-viewer.svelte:242]: I think you can use `previousMemory || previousAsset` here

File: web/src/lib/components/memory-page/memory-viewer.svelte [hunk 5]
```diff
@@ -246,29 +239,33 @@
...
+										<CircleIconButton
+											logo={ChevronLeft}
+											backgroundColor="#202123"
+											on:click={toPreviousAsset}
```
Comment (jrasm91) [inline web/src/lib/components/memory-page/memory-viewer.svelte:246]: Usually click methods use `handlePreviousAsset` or `onPreviousAsset` naming convention, depending if it is being handled or propagated.

File: web/src/lib/components/memory-page/memory-viewer.svelte [hunk 6]
```diff
@@ -42,119 +45,108 @@
...
+
+			toNextAsset();
+
+			if (!canAdvance()) {
+				pause();
+				resetStart();
+			}
+		};
+
+		const requestDraw = () => (animationFrameRequest = requestAnimationFrame(draw));
```
Comment (jrasm91) [inline web/src/lib/components/memory-page/memory-viewer.svelte:93]: It is a little strange to define this function here, reference it above (before the definition), and also it's implementation references `draw` which also calls it. I think it'd be possible to actually have the entire `draw` method be defined outside of the `play` method, and `play` only starts the animationFrameRequest loop if necessary.

**Review by jrasm91 — COMMENTED** [review]
We don't have one line if statements anywhere else in the codebase. In general, I think most of the developers (at least myself) prefer _not_ using if/for on single lines without braces.

Looks much cleaner though, thanks for adding those changes in. I think the next/current asset stuff as well.

File: web/src/lib/components/memory-page/memory-viewer.svelte (hunk 5 above)
Comment (uhthomas) [reply web/src/lib/components/memory-page/memory-viewer.svelte:246]: Makes sense, this is just what the original code was.

File: web/src/lib/components/memory-page/memory-viewer.svelte (hunk 6 above)
Comment (uhthomas) [reply web/src/lib/components/memory-page/memory-viewer.svelte:93]: Yeah, it is now. There was a point where I was working on it where `start` was defined in `play`. That's been moved out to handle pausing and navigation correctly, so I can now separate the function too.

**Review by jrasm91 — APPROVED** [review]

**Pull request merged**